### ISSUE TITLE
Question: Calling collapse from Rcpp

### ISSUE THREAD

**Issue by tony-aw** [SebKrantz/collapse#527]
Hi,

I have another short (and potentially stupid) question. Is it possible to call collapse C functions from Rcpp? If so, could you give a short working example? I tried it myself, but I failed. Thanks in advance.

Kind regards,

Tony.

**Comment by SebKrantz** [comment]
Hi, I haven't done this yet, but I think is possible. The C API is listed at the bottom of this file: https://github.com/SebKrantz/collapse/blob/master/src/ExportSymbols.c. So e.g. for fdist(), do trysomething like:
```
// perhaps put this in a header file
SEXP cp_dist(SEXP x, SEXP vec, SEXP Rret, SEXP Rnthreads);

 cp_dist = R_GetCCallable("collapse", "cp_dist");
```
Otherwise see Rcpp docs or contact Drik.

**Comment by eddelbuettel** [comment]
Seconded -- this should work, and we do this between some packages, while others (i.e. lme4 and Matrix) have done for many years.  The mechanics take a small moment of getting used to. The caller package needs to import the called package; that is needed as symbols gets exported / imported basically via R and not the system linker (via the `R_GetCCallable()` that @SebKrantz pointed to; there is a matching exporter in the other package).

My most recent (small) example is package crc32c providing a (C-level compiled) function (written in C++ but with a C interface) that package digest calls.  There are a few other packages doing it too.

It is actually all R mechanics; package Rcpp helps with interface generation (if you declare `Rcpp::interfaces(cpp)`; usually as `[[Rcpp::interfaces(r, cpp)]]` to also cover R) and with general conversion but the mechanism is all R.

**Comment by tony-aw** [comment]
I'm sorry, I'm still somewhat confused.
Where exactly should I add
```
// perhaps put this in a header file
SEXP cp_dist(SEXP x, SEXP vec, SEXP Rret, SEXP Rnthreads);

 cp_dist = R_GetCCallable("collapse", "cp_dist");
```
?

**Comment by eddelbuettel** [comment]
The above is how collapse exports a function `cp_dist`, declared as having four `SEXP` arguments and returning a `SEXP`, and `cp_dist` becomes the callable function of the same name from the `collapse` package.  You would use that in your code where `cp_dist()` is then the identifier for the four-parameter function made available by collapse.

Your best bet is probably to pick a (small enough) pair of packages that already do that.  I know we added an export to data.table years (see its file [datatableAPI.h](https://github.com/Rdatatable/data.table/blob/master/inst/include/datatableAPI.h) which we call in [package dtts](https://github.com/cran/dtts/blob/bdc04c248008339e03564982b4613a66c7da1a23/src/align.cpp#L331) -- this is all a bit complicated by the fact that `data.table` is moderately large and that there has been a discussion to renamed `subsetDT` as well). There are other pairs: `xts` exports a few functions but I now forget where I import them; `xts` if I recall also used some functions from `zoo` that way.

**Comment by SebKrantz** [comment]
Adding to this: I think basically should be able to add this to any C or C++ file in your package. Importing collapse should be sufficient. If not, LinkingTo might be necessary in the DESCRIPTIOn. You need to experiment around a bit.

**Comment by eddelbuettel** [comment]
Right. `LinkingTo` is common _if_ the exporting package defines a header declaring the accessible function(s) and doing the part where it gets those via `R_GetCCallable()` -- but that is not a requirement.   `Imports` is a requirement as the importing package needs access to the symbols and IIRC do a one-time assignment at load.

I think I once wrote myself a pair of very small demo packages.  I am a bit short on time but one day I may dig those out and make a tutorial as the question comes up every once in a while.

**Comment by tony-aw** [comment]
That would be very nice. I actually looked throught the documentation of Rcpp, but found very little practical examples of what I wanted, and I thought I had overlooked something crucial.

**Comment by eddelbuettel** [comment]
You are looking in the wrong place.  Try [Section 5.4.3 in Writing R Extensions](https://rstudio.github.io/r-manuals/r-exts/System-and-foreign-language-interfaces.html#linking-to-native-routines-in-other-packages) and study existing (working) examples.  It mentions `xts` and `zoo`.  

The question is really not related to `collapse`, or `Rcpp`.

**Comment by eddelbuettel** [comment]
Your simplest self-contained example may be in our package `dtts` in the C++ file `align.cpp` where `subsetDT` is the one exported `data.table` C-level function.  We call it with a data.frame / data.table object and row and column subsets.

There is not other setup.  `data.table` registers the routine 
https://github.com/Rdatatable/data.table/blob/cde7333938a590a3cddbda1b02103650e2f55d15/src/init.c#L172
which gets activated once you import `data.table` making the code in `dtts` work:
https://github.com/eddelbuettel/dtts/blob/9ce551173dae49554eaa1e56831258c512c75bcc/src/align.cpp#L279

**Comment by SebKrantz** [comment]
Thanks Dirk for the efforts. Unless there are specific issues related to the way the collapse C API is set up, I'll consider this issue as closed.

**Comment by tony-aw** [comment]
Hi,

I managed to do what I want, which is to call `collapse::whichv` from within Rcpp, to create a more efficient version of:
`lapply(v, \(i) collapse::whichv(x, i))`
where both `x` and `v` are vectors (in this case character vectors) containing occasional duplicate values.
Unfortunately, the simple 'Rcpp' implementation is actually SLOWER than the above lapply call. The difference is small, but it's still slower. So much for my effort :(.

I won't ask you for an explanation, as I think I understand why, and this isn't the 'Rcpp' repositiory anyway.

What I will ask is the following:
If it is not too much to ask, could you perhaps create a function equivalent to
`lapply(v, \(i) collapse::whichv(x, i))`,
but written in 'C'?
The reason is the above expression is useful, is that it is often needed to get all elements from an object where the object has duplicate names, and where the user can ask for duplicate indices, and where the order of the indices is relevant. And it would be nice to have something faster than `lapply()`.

Basically, `lapply(v, \(i) collapse::whichv(x, i))` would be a "match All()"-like function, except it accounts for both the order of `v` and `x`, as well as possible duplicate values.

Creating this function was surprisingly very easy in 'C++' (even without relying on `collapse::whichv`), but I don't have experience coding in 'C', and R's 'C++' API has some overhead. Hence my question to you. But I fully understand if you don't want to create such a function.

Thanks in advance.

Kind regards,

Tony

**Comment by SebKrantz** [comment]
How about this for a start?:

``` r
library(collapse)
#> collapse 2.0.9.9000, see ?`collapse-package` or ?`collapse-documentation`
#> 
#> Attaching package: 'collapse'
#> The following object is masked from 'package:stats':
#> 
#>     D
library(microbenchmark)

fmm <- function(x, v) {
  m <- fmatch(x, v)
  attr(m, "N.groups") <- length(v)
  oldClass(m) <- "qG"
  gsplit(g = m)
}

v <- letters
x <- sample(letters, 1e5, TRUE)

identical(lapply(v, whichv, x = x), fmm(x, v))
#> [1] TRUE

microbenchmark(
  lapply = lapply(v, whichv, x = x),
  fmm = fmm(x, v)
)
#> Unit: microseconds
#>    expr      min        lq     mean    median       uq      max neval
#>  lapply 2705.713 2731.9940 3475.943 2758.3570 2822.747 14754.10   100
#>     fmm  491.180  516.4975 1750.793  527.4445  555.427 36094.06   100
```

<sup>Created on 2024-01-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by tony-aw** [comment]
Hi,

First of all, thank you for the quick reply.
Maybe I'm missing something, but `fmm()` does not give the same result:

```
library(collapse)

n <- 1000
x <- sample(letters, n, TRUE) # contains duplicate values, and order matters
v <- sample(letters, n, TRUE) # contains duplicate values, and order matters

fmm <- function(x, v) {
  m <- fmatch(x, v)
  attr(m, "N.groups") <- length(v)
  oldClass(m) <- "qG"
  gsplit(g = m)
}
lapply(v, \(i) collapse::whichv(x, i))
fmm(x, v)

identical(lapply(v, \(i) collapse::whichv(x, i)), fmm(x, v))
```

`identical()` gives `FALSE` here. 

Some list elements are equal, and some are not equal. And it is not a matter of re-ordering. In fact, even the list lengths differ.
For example:

```
foo1 <- lapply(v, \(i) collapse::whichv(x, i))
foo2 <- fmm(x, v)

lengths(foo1) |> sort()
lengths(foo2) |> sort() # is different from foo1

```


Kind regards,

Tony

**Comment by SebKrantz** [comment]
Yes, you are correct in that `fmm()` will not work the same if there are duplicates in `v`.  So you either call `unique(v)` beforehand or you indeed need a multi match approach. I have added to my above comment the two implementations I know of. Both of them give vectorized output though (not a list of indices). I won't implement the list of indices because this is an expensive and inconvenient format to store the information. I can think about exporting `multi_match()`, or at lest making the C version available so you can turn the output into whatever you want.

**Comment by tony-aw** [comment]
Hi,

For transparency, and also as a thanks for taking your time answering my questions, here is my current solution. It required only a surprisingly small tweak to your code earlier (it turned out to be much, much simpler than I though):

```
match_all <- function(needles, haystack, unlist = TRUE) {
  v <- unique(needles)
  m <- collapse::fmatch(haystack, v)
  attr(m, "N.groups") <- length(v)
  oldClass(m) <- "qG"
  out <- collapse::gsplit(g = m)
  names(out) <- v
  if(unlist) return(unlist(out[needles], use.names = FALSE, recursive = FALSE))
  return(out[needles])
}
```

This works exactly like `lapply(v, \(i) collapse::whichv(x, i))`, allowing for duplicates in both vectors,
except it fulfills my unhealthy need for speed :-).

Thanks again, and kind regards,

Tony.


```

**Comment by SebKrantz** [comment]
If needles is unique you could add `if(length(needles) == length(v)) return(out)` to avoid the final subset (which may be expensive).
